### PR TITLE
pb-purge: Only delete files marked as "finished" or "deleted"

### PR DESCRIPTION
**Pull request by dennisschagt**
According to the documentation [1] and keymap [2], `pb-purge` only removes downloads marked as `finished` or `deleted`.
I think that makes sense but the actual implementation also removed downloads marked as `played`.

This PR fixes the inconsistency between implementation and documentation and adds a test for `QueueLoader::reload()`.
I'm still planning to write more tests for `QueueLoader` but that will be submitted in a separate Pull Request.

[1] https://github.com/newsboat/newsboat/blob/22bb5253773ec2b58213a4952db57426e852ba6c/doc/newsboat.asciidoc#user-content-pb-purge
[2] https://github.com/newsboat/newsboat/blob/22bb5253773ec2b58213a4952db57426e852ba6c/src/keymap.cpp#L203-L205

### CODE REVIEW

**Comment by Minoru** [comment]
`DlStatus` was always a bit of a mystery to me; I never invested the time to draw the state diagram and figure out what each state means exactly. So I'm not sure why we believe the docs, not the code, here.

As far as I remember, "finished" state is an explicit mark that the user sets on a download. Same goes for "deleted". On the other hand, "played" is set automatically after the podcast finished playing. Then it does make sense for the "purge" to only delete explicitly marked downloads, and leave "played" alone.

But another, perhaps more simplistic, standpoint is that "purge" should clear the queue of everything but the new, still-unprocessed downloads. Then it makes sense for the "purge" to also delete "played" items, as they've been heard by the user already.

Both viewpoints make sense to me, and I'm fine with picking either one; I just want to make sure I have the correct understanding, so e.g. I can explain it on IRC, and keep the same general line when adding other features.

**Comment by dennisschagt** [comment]
I created an overview so it can be easier to see how the different download states (DlStatus values) can be reached.
Might not be completely relevant to your question or this PR, I will try to explain my reasoning for this PR in the next comment.

image

| State              | queuefile text | Description   |
| ------------------ |:-------------- |:-------------:|
| QUEUED             | ""             | Default state when first added to the queue |
| DOWNLOADING        |                | Will automatically change to READY or FAILED or manually to CANCELLED (pb-cancel) |
| READY              | "downloaded"   | Ready to play (download was successful) |
| PLAYED             | "played"       | On pb-play after running `player` command. Return code of player is not checked |
| FINISHED           | "finished"     | Manually marked using pb-mark-as-finished |
| DELETED            |                | Will be removed from queuefile on exit or pb-purge |
| FAILED             |                | Download failed |
| CANCELLED          |                | Download was cancelled |
| ALREADY_DOWNLOADED |                | Default state if the podcast file (e.g. .mp3) already exists (otherwise, would be in QUEUED state) |

Some notes:
- `ALREADY_DOWNLOADED`: There can be multiple causes for the file to already exist , e.g.:
  - Downloaded before by podboat but it didn't save properly because of a crash or force-quit
  - File is created by some other application
- `DOWNLOADING`: Can be reached from any other state (using pb-download) but not shown in the diagram to keep it somewhat readable
- `PLAYED`: Switches to this state after the player returns independent of error code or how long it played.

**Comment by dennisschagt** [comment]
> As far as I remember, "finished" state is an explicit mark that the user sets on a download. Same goes for "deleted". 

Correct

> On the other hand, "played" is set automatically after the podcast finished playing.

"played" is set when the `player` program returns. Even if it returned with an error code or if the user quit the `player` program directly.

> Then it does make sense for the "purge" to only delete explicitly marked downloads, and leave "played" alone.
> 
> But another, perhaps more simplistic, standpoint is that "purge" should clear the queue of everything but the new, still-unprocessed downloads. Then it makes sense for the "purge" to also delete "played" items, as they've been heard by the user already.

I agree that that makes sense as well.

My main concern with the current behavior is that purge can easily make me lose a few entries.

For example, I might play a file for a few seconds, realize I don't have enough time (or get interupted) so I quit the player. Podboat than marks that entry as "played".
There is no way to change it back to "ready" except for downloading the file again (or manually editing the queue-file).

At a later time, if I want to purge for some reason, e.g.:
- if I want to remove all files which are marked as "finished"
- or just want to load new entries from the queue-file
I would lose the first file which I didn't finish listening to.

> Both viewpoints make sense to me, and I'm fine with picking either one; I just want to make sure I have the correct understanding, so e.g. I can explain it on IRC, and keep the same general line when adding other features.

I understand.

Keeping the old `pb-purge` behavior is fine with me but in that case I think it would be useful to add a `pb-reload` command so there is at least a way to load new entries from the queue-file without having to exit and restart podboat.

File: test/queueloader.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include <fstream>
+
+#include "3rd-party/catch.hpp"
+#include "configcontainer.h"
+#include "download.h"
+#include "test-helpers.h"
+
+using namespace podboat;
+
+bool contains_download_with_status(std::vector<Download>& downloads,
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply test/queueloader.cpp]: Oops, thanks for reminding me every time.
I might have to add a reminder in a local pre-commit as I'm still not consistently thinking about looking for `const` opportunities.

File: include/pbcontroller.h [hunk 2]
```diff
@@ -42,7 +42,7 @@ class PbController {
 	std::string get_formatstr();
 
 	unsigned int downloads_in_progress();
-	void reload_queue(bool remove_unplayed = false);
+	void reload_queue(bool also_remove_finished = false);
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply include/pbcontroller.h]: Good idea. Renamed the method to make clear that it is not just reloading the queue (`reload_queue()` -> `purge_queue()`)

**Comment by Minoru** [comment]
Thanks for the diagram and the user story! They made it all much more clear. Let's go with your proposed changes then.

**Comment by Minoru** [comment]
Merged via a917caf. Thanks for the work, @dennisschagt!

**Pull request closed without merging**